Use native node crypto

aljones15 opened the pull request:
Features in this PR:

1. adds a node and browser specific `ed25519` file.
2. the node specific ed25519 file uses node's crypto `sign`, `verify`, `createPrivateKey` and `createPublicKey`
3. the browser specific ed25519 uses `@stablelib/ed25519`
4. the node specific ed25519 is able to turn Uint8Arrays and or Buffers into DER encoded keys for node


Addresses:
https://github.com/digitalbazaar/ed25519-verification-key-2020/issues/2

Suggestions:

We really should create a test suite that ensures that the node ed25519 keys, can in turn verify stablelib's keys, and possibly node-forge's keys.

A [sister PR](https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/4) removes webpack's node utils reducing the library's size to 33kb.

aljones15 committed:
Use crypto interface across browser and node.

aljones15 committed:
Add utils from 2018 key to enable node 25519 keys.

aljones15 committed:
Use add ported interfaces for node ed25519.

aljones15 committed:
Add publicBytes to secretBytes.

aljones15 committed:
Use DER encodings for verify and sign.

aljones15 commented on package.json:
```diff
@@ -18,7 +18,8 @@
     "crypto-ld": "^4.0.2",
-    "esm": "^3.2.25"
+    "esm": "^3.2.25",
+    "node-forge": "^0.10.0"
```
how do I ensure that node-forge is not loaded in the browser? An alias correct? EDIT: looks like this is an `external` property

aljones15 replied:
used an external and it works

aljones15 commented on lib/crypto.js:
```diff
@@ -0,0 +1,86 @@
+
+// FIXME someone more familar with TypedArrays and Buffers
+// should review this
+const typedArrayToBuffer = array => {
```
@dlongley @gannan08  does this look correct to you?

aljones15 replied:
Addressed this here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/7efbbfd8d01fe357a5593ec4095533ea8d108c96

aljones15 commented on lib/crypto.js:
```diff
@@ -0,0 +1,86 @@
+      asn1.fromDer(new ByteBuffer(privateKey.export(privateKeyEncoding))));
+    return {
+      publicKey: publicKeyBytes,
+      secretKey: Buffer.concat([privateKeyBytes, publicKeyBytes])
```
@dlongley @gannan08 @mattcollier  is concatenating 2 Buffers together always going to result in the correct buffer?

dlongley replied:
That's how `secretKey` is formatted in NACL/etc. -- 32 bytes of private key followed by 32 bytes of public key.

aljones15 committed:
Add webpack tools to devDeps.

aljones15 committed:
Remove node forge (#5)

* Use DER Buffer over node-forge.

* Remove forge from public key der encode.

* Remove node-forge.

* Use KEY_PREFIX at top of file.

* Use hexstring to save space in PUBLIC DER PREFIX.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use slice over getting ArrayBuffer.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use hex string from PRIVATE DER PREFIX.

* Add zeroBuffer to PUBLIC DER PREFIX.

* Use `instanceof Uint8Array` instead of `Buffer.isBuffer`.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

* Use instanceof Uint8Array instead of isBuffer for private key material.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Add github actions and node-forge again.

dmitrizagidulin commented on lib/util.js:
```diff
@@ -0,0 +1,74 @@
+ * Copyright (c) 2018-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+
+const DER_PRIVATE_KEY_PREFIX = Buffer.from(
```
We should move the contents of this `util.js` file into the Node-only `lib/crypto.js` file. (Since it should be left out of the browser build).

aljones15 replied:
probably, I'm looking into one further optimization, I'm not sure sure we need to call on node `createPrivateKey` and `createPublicKey` in gen key as it appears all we need are the Buffers containing the key material. We do still call on those functions in sign and verify.

aljones15 replied:
@dmitrizagidulin addressed this here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/2db9b78f875d45c9debe228af2ba0d7f6b049082 and https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/267da57447389a809150bcbd2507a0f77f33ee70

p.s. it turns out we do need the node key creation in generate.

aljones15 committed:
Remove node-forge from crypto.js too.

aljones15 committed:
Move node specific DER stuff to crypto.js.

aljones15 committed:
Delete lib util.js.

aljones15 committed:
Add jsdoc strings to DerEncode functions.

aljones15 committed:
Further comment comment node crypto file.

aljones15 committed:
Add copyrights to new files.

dlongley commented on lib/Ed25519VerificationKey2020.js:
```diff
@@ -2,7 +2,7 @@
  */
 import * as base58btc from 'base58-universal';
-import * as ed25519 from '@stablelib/ed25519';
+import crypto from './crypto';
```
-1 to naming a module `crypto` unless it uses a commonly expected crypto API (node.js or WebCrypto). This seems like it's some ed25519 lib, so we could do `ed25519.js` unless you've got a better name than that.

aljones15 replied:
I was using `key` in another branch.

dlongley commented on lib/Ed25519VerificationKey2020.js:
```diff
@@ -2,7 +2,7 @@
-import * as ed25519 from '@stablelib/ed25519';
+import crypto from './crypto';
 import {LDVerifierKeyPair} from 'crypto-ld';
 import {randomBytes} from '@stablelib/random';
```
Can we hide this detail inside the browser version file that implements the crypto APIs?

aljones15 replied:
I like this idea, you're a mad man @dlongley 
I'll move {randomBytes} inside whatever we rename crypto to.

dlongley replied:
Let's rename `key` to `ed25519`.

aljones15 replied:
done: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/202ce82d3b3839738563ededf6a6da865918c50a

dlongley replied:
Thanks! Side Note: Still need to move `randomBytes` inside as an implementation detail though.

dlongley commented on lib/Ed25519VerificationKey2020.js:
```diff
@@ -73,16 +73,13 @@ export class Ed25519VerificationKey2020 extends LDVerifierKeyPair {
-      keyObject = ed25519.generateKeyPair({
-        isAvailable: true,
+      keyObject = await crypto.generateKeyPair({
         randomBytes: secureRandom
```
I'd rather `secureRandom` not even be an option above as a method parameter. Can we remove it -- it's a footgun. Why do we have it?

aljones15 replied:
I don't know why we have this, but if you want to make a suggestion for how to fix this I'll implement it and we can get this PR done tonight.

dlongley replied:
The use of this lib should not be able to pass in a `secureRandom` anything, IMO. We should do this internally within `ed25519.js`. The node.js crypto version doesn't need it at all and should use `crypto.randomBytes` if it needs to generate a seed. The browser version should use the built-in from stable-lib that relies on WebCrypto to function. If neither of those is available, there should be an error that's generated (hopefully) by stable-lib. But this module is supposed to help prevent people from shooting themselves in the foot, so we shouldn't give them and option to pick something that's not secure.

If the stable-lib implementation requires us to pick something, then update the `ed25519.js` for the browser version and put the `secureRandom` import in there directly in the key gen function. Don't surface it as an option on the public API.

aljones15 replied:
Ok so the easiest way to implement this was to remove `secureRandom` as a parameter. Also it turns out that stable lib needs `randomBytes` to be a function.

https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/2f5c06125b12a7ed9fe79bd8d73e07e48a427192

dlongley replied:
@aljones15 -- actually, it's probably better to just call `crypto.getRandomValues` in the browser version and not import `secureRandom` at all, we don't need the extra code.

aljones15 replied:
cool did that here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/11d6eb9cf0ce47777e6fe4ec4e6386a922b94ca5

moved randomBytes into the browser and node stuff and removed the export.

dlongley commented on lib/crypto-browser.js:
```diff
@@ -0,0 +1,7 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+
```
```suggestion
```

dlongley commented on lib/crypto.js:
```diff
@@ -0,0 +1,166 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+
```
```suggestion
```

dlongley commented on lib/crypto.js:
```diff
@@ -0,0 +1,166 @@
+ * @returns {Buffer} The key material part of the Buffer.
+*/
+function getKeyMaterial(buffer) {
+  if(buffer.indexOf(DER_PUBLIC_KEY_PREFIX, 0) === 0) {
```
```suggestion
  if(buffer.indexOf(DER_PUBLIC_KEY_PREFIX) === 0) {
```

aljones15 replied:
while unlikely, I think it is better to start at index 0 to ensure that the public key material after the prefix is not somehow a match for the prefix itself.

dlongley commented on lib/crypto.js:
```diff
@@ -0,0 +1,166 @@
+  if(buffer.indexOf(DER_PUBLIC_KEY_PREFIX, 0) === 0) {
+    return buffer.slice(DER_PUBLIC_KEY_PREFIX.length, buffer.length);
+  }
+  if(buffer.indexOf(DER_PRIVATE_KEY_PREFIX, 0) === 0) {
```
```suggestion
  if(buffer.indexOf(DER_PRIVATE_KEY_PREFIX) === 0) {
```

aljones15 replied:
same as above, I think it's better to specify 0 here so that if some combination of the prefix and the private key material happens to match the prefix we don't end up matching that.

dlongley replied:
I think there's confusion over what [the API](https://nodejs.org/api/buffer.html#buffer_buf_indexof_value_byteoffset_encoding) does. Passing `0` or not passing `0` does the same thing, as `0` is the default value for the second parameter. It is the check `=== 0` that ensures that the prefix is at the start of the buffer.

dlongley commented on lib/crypto.js:
```diff
@@ -0,0 +1,166 @@
+
+// FIXME someone more familar with TypedArrays and Buffers
+// should review this
+const typedArrayToBuffer = array => {
+  const length = array.byteLength + array.byteOffset;
+  const arrayBuffer = array.buffer.slice(array.byteOffset, length);
+  return Buffer.from(arrayBuffer, array.byteOffset, length);
+};
```
All you need is: `Buffer.from(typedArrayHere);`. So we should remove this function and just use this where it's needed -- if it's even needed at all.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/7efbbfd8d01fe357a5593ec4095533ea8d108c96 and thanks for clarifying this.

dlongley commented on lib/crypto.js:
```diff
@@ -0,0 +1,166 @@
+    const privateKeyMaterial = privateBytesToDer(privateBytes);
+    // node is more than happy to create a new private key using a DER
+    const privateKey = await createPrivateKey({
+      key: privateKeyMaterial,
```
```suggestion
      key: Buffer.from(privateBytes),
```

aljones15 replied:
this will not work not, Node's `crypto.createPrivateKey` does require the DER prefix before the private bytes.

dlongley replied:
Sorry, my suggestion should have been:

```suggestion
      key: privateKeyToDer(privateBytes),
```

I was trying to get rid of the extra variable and keep this consistent with what I was expecting to be other changes (haven't looked yet).

aljones15 replied:
https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/2c08aba77dd75a8d202a01667def91e0164fb168

you were off by a few commits in the method name, so I made the commit above and removed `privateKeyMaterial`.

dlongley commented on lib/crypto.js:
```diff
@@ -0,0 +1,166 @@
+    const privateBytes = new Uint8Array(64);
+    privateBytes.set(seedBytes);
+    privateBytes.set(publicBytes, 32);
+    const privateKeyMaterial = privateBytesToDer(privateBytes);
```
```suggestion
```

aljones15 replied:
ok so the issue here is that node's `crypto.createPrivateKey` needs the DER prefix in order to figure out the type of key it's creating. On the other hand though we could do something like:

```js
Buffer.concat([DER_PRIVATE_KEY_PREFIX, Buffer.from(privateBytes)]);
```

and remove the call to `privateKeyDerEncode` in this function (we still need it for sign though).

aljones15 replied:
this should be address here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/2c08aba77dd75a8d202a01667def91e0164fb168

removed `privateKeyMaterial`

dlongley commented on lib/crypto.js:
```diff
@@ -0,0 +1,166 @@
+  return seed;
+};
+
+// helper function that takes an TypedArray of private bytes and turns it
+// into a Buffer containing a DER encoded key
+const privateBytesToDer = array => privateKeyDerEncode(
+  {privateKeyBytes: typedArrayToBuffer(array)});
+
+// helper function that takes a TypedArray of public bytes and turns it
+// into a Buffer containing a DER encoded key
+const publicBytesToDer = array => publicKeyDerEncode(
+  {publicKeyBytes: typedArrayToBuffer(array)});
```
Can we remove these entirely and just call `privateKeyDerEncode` and `publicKeyDerEncode` directly where needed? We may want to use the names `privateKeyToDer` and `publicKeyToDer` instead though.

aljones15 replied:
Addressed here and good idea: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/7efbbfd8d01fe357a5593ec4095533ea8d108c96

dlongley reviewed (changes requested):
There are a number of superfluous encoding/translation things that seem to be happening. Let's see if we can tighten it up. See comments.

aljones15 committed:
Remove empty line from crypto-browser.js.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove empty line from crypto.js.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove typedArrayToBuffer & add Buffer.from to ToDer methods.

aljones15 committed:
Rename crypto to ed25519 and use key at the module name.

aljones15 committed:
Use buffer.indexOf default value of 0.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use buffer.indexOf default of 0.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove privateKeyMaterial var.

dlongley commented on lib/Ed25519VerificationKey2020.js:
```diff
@@ -2,7 +2,7 @@
  */
 import * as base58btc from 'base58-universal';
-import * as ed25519 from '@stablelib/ed25519';
+import key from './ed25519';
```
```suggestion
import ed25519 from './ed25519.js';
```

aljones15 replied:
renaming from `key` to `ed25519` would have broken the CI so did the commit with all the changes here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/202ce82d3b3839738563ededf6a6da865918c50a

aljones15 committed:
Use ed25519 over key as module name.

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,146 @@
+}
+
+/**
+ * Takes a Buffer or Uint8Array and adds a DER Private Prefix to it.
+ * Allows Uint8Arrays to be interoperable with node's crypto functions.
```
This comment is too implementation specific -- what it is really doing is encoding the raw private key in DER-encoded PKCS#8 format.

```suggestion
 * Takes a Buffer or Uint8Array with the raw private key and encodes it
 * in DER-encoded PKCS#8 format.
 * Allows Uint8Arrays to be interoperable with node's crypto functions.
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,146 @@
+}
+
+/**
+ * Takes a Buffer of public key bytes and adds the DER Public Key Prefix to it.
```
```suggestion
 * Takes a Buffer of public key bytes and encodes it in DER-encoded SubjectPublicKeyInfo (SPKI) format.
```

dlongley commented on lib/Ed25519VerificationKey2020.js:
```diff
@@ -2,7 +2,7 @@
  */
 import * as base58btc from 'base58-universal';
-import * as ed25519 from '@stablelib/ed25519';
+import ed25519 from './ed25519';
```
Always include the extension if it is a file and not a module that is being imported:

```suggestion
import ed25519 from './ed25519.js';
```

aljones15 committed:
Make jsdoc description for privateKeyDerEncode less implementation specific.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,146 @@
+    // extract the first 32 bytes of the 64 byte private key representation
+    p = privateKeyBytes.slice(0, 32);
+  }
+  return Buffer.concat([DER_PRIVATE_KEY_PREFIX, Buffer.from(p)]);
```
`Buffer.concat` accepts `Uint8Array`s so we don't need the extra `Buffer.from` here or any other place.

```suggestion
  return Buffer.concat([DER_PRIVATE_KEY_PREFIX, p]);
```

aljones15 replied:
There were two instances of this corrected here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/93e0432449d939674f819f917bf7279295309ba5

aljones15 committed:
Expand jsdoc comment to include format.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Include file extension in es6 import.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove Buffer.from from DerEncoding methods.

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,147 @@
+  async generateKeyPairFromSeed(seed) {
+    const seedBytes = await getSeedBytes(seed);
+    const publicBytes = new Uint8Array(32);
+    const privateBytes = new Uint8Array(64);
+    privateBytes.set(seedBytes);
+    privateBytes.set(publicBytes, 32);
```
This `privateBytes` code is confusing. If we have `seedBytes` we should be passing those through to `privateKeyDerEncode`.

```suggestion
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,147 @@
+    privateBytes.set(publicBytes, 32);
+    const privateKey = await createPrivateKey({
+      // node is more than happy to create a new private key using a DER
+      key: privateKeyDerEncode({privateKeyBytes: privateBytes}),
```
```suggestion
      key: privateKeyDerEncode({seedBytes}),
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,147 @@
+    // this expects either a PEM encoded key or a node privateKeyObject
+    const publicKey = await createPublicKey(privateKey);
+    const publicKeyBuffer = publicKey.export(publicKeyEncoding);
+    const privateKeyBuffer = privateKey.export(privateKeyEncoding);
+    const publicKeyBytes = getKeyMaterial(publicKeyBuffer);
+    const privateKeyBytes = getKeyMaterial(privateKeyBuffer);
+    return {
+      publicKey: publicKeyBytes,
+      secretKey: Buffer.concat([privateKeyBytes, publicKeyBytes])
```
We don't need to export the private key at all, right? We can just use the `seedBytes` below?

```suggestion
    const publicKey = getKeyMaterial(publicKeyBuffer);
    return {
      publicKey,
      secretKey: Buffer.concat([seedBytes, publicKeyBytes])
```

aljones15 replied:
A few other variables had to be removed in order to remove this the commit is here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/fa7a27d0bb20a17a023edc036c69f33a4e981229

dlongley replied:
Yay for simplification! :)

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,147 @@
+  async generateKeyPair({randomBytes}) {
+    return api.generateKeyPairFromSeed(randomBytes);
+  },
+  async sign(privateKeyBuffer, data) {
+    const privateKey = await createPrivateKey({
+      key: privateKeyDerEncode({privateKeyBytes: privateKeyBuffer}),
```
```suggestion
  async sign(privateKeyBytes, data) {
    const privateKey = await createPrivateKey({
      key: privateKeyDerEncode({privateKeyBytes}),
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,147 @@
+    });
+    return sign(null, data, privateKey);
+  },
+  async verify(publicKeyBuffer, data, signature) {
+    const publicKey = await createPublicKey({
+      key: publicKeyDerEncode({publicKeyBytes: publicKeyBuffer}),
```
```suggestion
  async verify(publicKeyBytes, data, signature) {
    const publicKey = await createPublicKey({
      key: publicKeyDerEncode({publicKeyBytes}),
```

aljones15 committed:
Move randomBytes into ed25519 files.

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,148 @@
+  return Buffer.concat([DER_PUBLIC_KEY_PREFIX, publicKeyBytes]);
+}
+
+// seedBytes can be a TypedArray or a function that returns a TypedArray
+const getSeedBytes = async seed => {
+  if(typeof seed === 'function') {
+    return seed();
+  }
+  return seed;
+};
```
```suggestion
```

aljones15 replied:
we do need this as `stablelib` requires that `randomBytes` is a function.

dlongley replied:
We're in the node version that doesn't use stablelib here.

aljones15 replied:
Removed getSeedBytes here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/11d6eb9cf0ce47777e6fe4ec4e6386a922b94ca5

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,148 @@
+
+const api = {
+  async generateKeyPairFromSeed(seed) {
+    const seedBytes = await getSeedBytes(seed);
```
We shouldn't call `getSeedBytes` but instead validate that `seed` is a 32-byte Uint8Array.

aljones15 replied:
in order to be backwards compatible with `stablelib/ed25519` we do. It might be better to accept that seed is a function if we want to simplify.

dlongley replied:
We're miscommunicating. First, we don't need to be "backwards compatible" with `stablelib/ed25519` because it is just an implementation detail. Second, we should change the internal module interface that is split between the browser/node to do what we need; we don't have to code directly to `stablelib/ed25519`'s API if that causes some trouble, which it seems to be doing here -- it's spilling details that should be hidden in the browser version into the code that is used in both the browser and node.

aljones15 replied:
I believe this address this issue: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/11d6eb9cf0ce47777e6fe4ec4e6386a922b94ca5

aljones15 committed:
Remove secureRandom.

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,148 @@
+      secretKey: Buffer.concat([privateKeyBytes, publicKeyBytes])
+    };
+  },
+  async generateKeyPair({randomBytes}) {
+    return api.generateKeyPairFromSeed(randomBytes);
```
Somewhere at the top of the file add:

```js
const generateRandomBytes = promisify(crypto.randomBytes);
```

```suggestion
  async generateKeyPair() {
    const seed = await generateRandomBytes(32);
    return api.generateKeyPairFromSeed(seed);
```

aljones15 replied:
node's `crypto.randomBytes` returns a Buffer, so no need to the promisify it looks like:

https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/c0ad8029f48bcff17eb8e19701b01a4b506a895d

dlongley replied:
@aljones15 -- We're miscommunicating here. It returns a Buffer because it's synchronous. That means if the PRNG stalls (due to lack of entropy), you'll block the whole event loop (whole app) waiting on it. Now, I don't think there are any/many node.js PRNG implementations that don't use something like Fortuna that are able to avoid that in probably all cases. But it's something I've been burned by in the past and that's what I mention it. We're fine as is and we will promisify it later if it becomes a real problem.

dlongley replied:
I see you promisified in another commit. Resolving this. Thanks!

aljones15 committed:
Ensure randomBytes is a function.

aljones15 committed:
Use seedBytes over privateKeyBytes.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove privateBytes from generate.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Change parameter of verify to match parameters of publicKeyDerEncode.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Change parameters of sign to match privateKeyDerEncode.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use randomBytes in generateKeyPair method.

dlongley commented on lib/ed25519-browser.js:
```diff
@@ -0,0 +1,11 @@
+export default {
+  // use stablelib's randomBytes in the browser
+  randomBytes,
+  ...ed25519
```
Instead of exposing `randomBytes` here, let's do:

```suggestion
  ...ed25519,
  generateKeyPair
```

And define `generateKeyPair` to call `ed25519.generateKeyPair` and use `randomBytes` internally.

aljones15 replied:
Addressed this here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/11d6eb9cf0ce47777e6fe4ec4e6386a922b94ca5

dlongley commented on lib/ed25519-browser.js:
```diff
@@ -0,0 +1,11 @@
+import {randomBytes} from '@stablelib/random';
+
+export default {
+  // use stablelib's randomBytes in the browser
+  randomBytes,
```
```suggestion
```

aljones15 replied:
removed randomBytes as an export here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/11d6eb9cf0ce47777e6fe4ec4e6386a922b94ca5

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,147 @@
+  sign,
+  verify,
+  createPrivateKey,
+  createPublicKey,
+  randomBytes
```
We don't need to expose `randomBytes`. That's only being done because stablelib needs it to be passed into its `generateKeyPair` function. But that's an implementation detail of the browser version that has no business influencing this version -- this is breaking abstraction. What we need to do is expose functions in our common `ed25519` API that do what we need (and only the common behavior that doesn't let people pass in PRNGs) ... and hide the other details behind the scenes.

aljones15 replied:
I believe this address this issue: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/11d6eb9cf0ce47777e6fe4ec4e6386a922b94ca5

aljones15 committed:
Move randomBytes into respective env specific modules.

aljones15 committed:
Remove export of privateKeyBuffer in favor of using seedBytes.

dlongley commented on lib/Ed25519VerificationKey2020.js:
```diff
@@ -60,29 +59,17 @@ export class Ed25519VerificationKey2020 extends LDVerifierKeyPair {
    * @param {object} [options={}] - Options hashmap.
    * @param {Uint8Array|Buffer} [options.seed] - A 32-byte array seed for a
```
A Buffer is an instance of Uint8Array.

```suggestion
   * @param {Uint8Array} [options.seed] - A 32-byte array seed for a
```

dlongley commented on lib/ed25519-browser.js:
```diff
@@ -0,0 +1,12 @@
+import * as ed25519 from '@stablelib/ed25519';
+import {randomBytes} from '@stablelib/random';
+
+export default {
+  ...ed25519,
+  async generateKeyPair() {
+    return ed25519.generateKeyPair({randomBytes: () => randomBytes(32)});
+  }
+};
```
No need to create a new function every time we generate a key pair:

```suggestion
export default {
  ...ed25519,
  async generateKeyPair() {
    const seed = await randomBytes(32);
    return ed25519.generateKeyPairFromSeed(seed);
  }
};
```

aljones15 replied:
`@stablelib/ed25519` needs a function. I did create one that is re-usable that can be passed in here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/0fcf12662180890a3fdce649b03583a089b738fc

NOTE: this is no longer the case we now use `generateKeyPairFromSeed`: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/b3c0b03ea9a1638c11aafd12a8b1a2e499533945

dlongley commented on lib/ed25519-browser.js:
```diff
@@ -0,0 +1,12 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+import * as ed25519 from '@stablelib/ed25519';
+import {randomBytes} from '@stablelib/random';
+
+export default {
+  ...ed25519,
+  async generateKeyPair() {
+    return ed25519.generateKeyPair({randomBytes: () => randomBytes(32)});
+  }
+};
```
No need to create a new function every time we generate a key pair -- and we don't need the extra library:

```suggestion
/*!
 * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
 */
import * as ed25519 from '@stablelib/ed25519';

// browser MUST provide "crypto.getRandomValues"
const crypto = self && (self.crypto || self.msCrypto);
if(!crypto.getRandomValues) {
  throw new Error('Browser does not provide "crypto.getRandomValues".');
}

export default {
  ...ed25519,
  async generateKeyPair() {
    const seed = new Uint8Array(32);
    crypto.getRandomValues(seed);
    return ed25519.generateKeyPairFromSeed(seed);
  }
};
```

aljones15 replied:
ok switched to using `generateKeyPairFromSeed` and removed the import of `@stabelib/random`: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/b3c0b03ea9a1638c11aafd12a8b1a2e499533945

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,140 @@
+function privateKeyDerEncode({privateKeyBytes, seedBytes}) {
+  if(!(privateKeyBytes || seedBytes)) {
+    throw new TypeError('`privateKeyBytes` or `seedBytes` is required.');
+  }
+  if(!privateKeyBytes && !(seedBytes instanceof Uint8Array &&
+    seedBytes.length === 32)) {
+    throw new TypeError('`seedBytes` must be a 32 byte Buffer.');
+  }
+  if(!seedBytes && !(privateKeyBytes instanceof Uint8Array &&
+    privateKeyBytes.length === 64)) {
+    throw new TypeError('`privateKeyBytes` must be a 64 byte Buffer.');
+  }
+  let p;
+  if(seedBytes) {
+    p = seedBytes;
+  } else {
+    // extract the first 32 bytes of the 64 byte private key representation
+    p = privateKeyBytes.slice(0, 32);
+  }
+  return Buffer.concat([DER_PRIVATE_KEY_PREFIX, p]);
+}
+
+/**
+ * Takes a Buffer of public key bytes and encodes it in DER-encoded
+ * SubjectPublicKeyInfo (SPKI) format.
+ * Allows Uint8Arrays to be interoperable with node's crypto functions.
+ *
+ * @param {object} options - Options to use.
+ * @param {Uint8Array} options.publicKeyBytes - The keyBytes.
+ *
+ * @throws {TypeError} Throws if the bytes are not Uint8Array or of length 32.
+ *
+ * @returns {Buffer} DER Public key Prefix + key bytes.
+*/
+function publicKeyDerEncode({publicKeyBytes}) {
+  if(!(publicKeyBytes instanceof Uint8Array && publicKeyBytes.length === 32)) {
+    throw new TypeError('`publicKeyBytes` must be a 32 byte Buffer.');
+  }
+  return Buffer.concat([DER_PUBLIC_KEY_PREFIX, publicKeyBytes]);
+}
```
Style fix: Move helper functions below public API.

aljones15 replied:
I moved them below the definition of `const = api {` but above the `export default api`.

https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/0fcf12662180890a3fdce649b03583a089b738fc

is that ok?

dlongley replied:
I think we should just move the export line up too and simplify further.

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,140 @@
+
+const api = {
+  /**
+   * Generates a key using a 32 bit Uint8Array.
```
```suggestion
   * Generates a key using a 32 byte Uint8Array.
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,140 @@
+   *
+   * @param {Uint8Array} seedBytes - The bytes for the private key.
+   *
+   * @returns {object} The buffer for the public and private key material.
```
```suggestion
   * @returns {object} The object with the public and private key material.
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,140 @@
+  },
+  // generates an ed25519 key using a random seed
+  async generateKeyPair() {
+    return api.generateKeyPairFromSeed(randomBytes(32));
```
Note: This will use the synchronous version of `randomBytes`. I was suggesting we use the asynchronous version (which requires promisifying it) -- but we can change to that later if it's really an issue. I won't hold this PR up over the difference -- and we don't have data indicating which would be better in a given system anyway. It isn't likely to have a significant event loop blocking impact (I don't think node.js's PRNG implementation ever blocks for significant periods of time).

aljones15 replied:
if you can link me to the promisify version that is fine, but as far I can tell it now returns a promise it looks like.

EDIT: my apologies. I can't seem to find a version that required a callback that is in turn promisifyable.

this version return sync.

EDIT:

ok so the docs say if a callback is passed in `randomBytes` is async, `promisify` should pass in a callback there for this should be an async version of `randomBytes`: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/4f0da5a8cff8bb2f43c5d77d71cddb8da8c1d421

dlongley replied:
Yes, this is what I was suggesting we do to ensure we don't block the event loop.

dlongley reviewed (changes requested):
Just a few more changes, getting close!

aljones15 committed:
Use Uint8Array as jsdoc type over Buffer.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use byte over bit in jsdoc doc strings.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Clarify jsdoc return type for ed25519 key.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Move helpers function below api defs.

  - Create a helper function for randomBytes for @stablelib.
  - Use variable for seed in node.

aljones15 committed:
Use window.crypto.getRandomValues to generate seed to @stabelib ed25519.

aljones15 committed:
Promisify randomBytes to make async.

dlongley commented on lib/ed25519-browser.js:
```diff
@@ -0,0 +1,24 @@
+  throw new Error('Browser does not provide "crypto.getRandomValues".');
+}
+
+// helper function used by stablelib to generate seeds
+const randomBytes = () => {
+  const seed = new Uint8Array(32);
+  return crypto.getRandomValues(seed);
+};
+
+export default {
+  ...ed25519,
+  async generateKeyPair() {
+    const seed = randomBytes();
```
This function isn't used by stablelib, we're just using it here to generate seeds. Also, it's not a generic `randomBytes` function, rather it specifically generates 32 random bytes for a seed. That means it should be renamed `generateRandomSeed`. But I don't think we need the complexity of the function at all (I'd just put the two lines of code directly into the function below:

```suggestion
export default {
  ...ed25519,
  async generateKeyPair() {
    const seed = new Uint8Array(32);
    crypto.getRandomValues(seed);
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,145 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+import {
+  sign,
+  verify,
+  createPrivateKey,
+  createPublicKey,
+  randomBytes
+} from 'crypto';
+
+const {promisify} = require('util');
+
+const randomBytesAsync = promisify(randomBytes);
```
Don't use `require` (use ESM import):

```suggestion
import {
  sign,
  verify,
  createPrivateKey,
  createPublicKey,
  randomBytes
} from 'crypto';
import promisify from 'util';

const randomBytesAsync = promisify(randomBytes);
```

aljones15 replied:
resolved this here: https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/07ee808ebc5569c235809a7d7645b9b18a26fef4

You did need

```js
import {promisify} from 'util'
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,145 @@
+  }
+  return Buffer.concat([DER_PUBLIC_KEY_PREFIX, publicKeyBytes]);
+}
+
+export default api;
```
```suggestion
```

dlongley commented on lib/ed25519.js:
```diff
@@ -0,0 +1,145 @@
+// used to turn public key bytes into a buffer in DER format
+const DER_PUBLIC_KEY_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');
+
+const api = {
```
```suggestion
export default {
```

dlongley reviewed (approved):
At this point there are just a few nits to get cleaned up so I'm going to approve assuming those will go in. Thanks!

aljones15 committed:
Move Export line up.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Export object as default.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Remove helper function in favor of direct crypto call.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Use import on util & move export default api up.

aljones15 committed:
Correct documentation to use Uint8Array over Buffer.
